File: cmd/uninstall.go
package cmd

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/mturley/agent-handler/db"
	watcherPkg "github.com/mturley/agent-handler/watcher"
	"github.com/spf13/cobra"
)

var uninstallCmd = &cobra.Command{
	Use:   "uninstall",
	Short: "Remove agent-handler configuration",
	RunE:  runUninstall,
}

func init() {
	uninstallCmd.GroupID = "admin"
	rootCmd.AddCommand(uninstallCmd)
}

var skillNames = []string{
	"inbox",
	"inbox-clear",
	"inbox-mode",
	"message",
	"watch",
	"unwatch",
	"watching",
	"handler",
	"catchup",
	"done",
	"block",
	"unblock",
	"reminder",
	"handler-debug",
	"ui",
}

func runUninstall(cmd *cobra.Command, args []string) error {
	home, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("failed to get home directory: %w", err)
	}

	claudeDirs, err := selectClaudeDirs(home, false)
	if err != nil {
		return err
	}

	agentHandlerDir := db.HandlerHome()

	// Summarize what will be done
	fmt.Println("agent-handler uninstall will:")
	fmt.Println("")

	fmt.Printf("  Clean up %d Claude Code configuration director%s:\n", len(claudeDirs), pluralIES(len(claudeDirs)))
	for _, cd := range claudeDirs {
		fmt.Printf("    - %s\n", cd)
	}

	type dirTargets struct {
		dir             string
		claudeSkillsDir string
		settingsPath    string
		claudeRulesDir  string
		symlinkTargets  []string
		hookNames       []string
		installedRules  []string
		hasPermission   bool
	}
	var perDir []dirTargets

	for _, cd := range claudeDirs {
		claudeSkillsDir := filepath.Join(cd, "skills")
		settingsPath := filepath.Join(cd, "settings.json")
		claudeRulesDir := filepath.Join(cd, "rules")

		symlinkTargets := findAgentHandlerSkills(claudeSkillsDir)
		hookNames := findAgentHandlerHooks(settingsPath)
		hasPermission := hasHandlerPermission(settingsPath)

		var installedRules []string
		rulesPath := filepath.Join(agentHandlerDir, "rules")
		if entries, err := os.ReadDir(rulesPath); err == nil {
			for _, entry := range entries {
				ruleDst := filepath.Join(claudeRulesDir, entry.Name())
				if _, err := os.Stat(ruleDst); err == nil {
					installedRules = append(installedRules, entry.Name())
				}
			}
		}

		fmt.Printf("\n  In %s:\n", cd)
		if len(symlinkTargets) > 0 {
			fmt.Printf("    Remove %d skill symlinks from %s:\n", len(symlinkTargets), claudeSkillsDir)
			for _, name := range symlinkTargets {
				fmt.Printf("      - %s\n", name)
			}
		}
		if len(hookNames) > 0 {
			fmt.Printf("    Remove %d hooks from %s:\n", len(hookNames), settingsPath)
			for _, name := range hookNames {
				fmt.Printf("      - %s\n", name)
			}
		}
		if len(installedRules) > 0 {
			fmt.Printf("    Remove %d global rule(s) from %s:\n", len(installedRules), claudeRulesDir)
			for _, name := range installedRules {
				fmt.Printf("      - %s\n", name)
			}
		}
		if hasPermission {
			fmt.Printf("    Remove Bash(handler *) permission from %s\n", settingsPath)
		}

		perDir = append(perDir, dirTargets{
			dir:             cd,
			claudeSkillsDir: claudeSkillsDir,
			settingsPath:    settingsPath,
			claudeRulesDir:  claudeRulesDir,
			symlinkTargets:  symlinkTargets,
			hookNames:       hookNames,
			installedRules:  installedRules,
			hasPermission:   hasPermission,
		})
	}
	fmt.Println("")

	// Check cmux actions
	cmuxActionsPresent := hasCmuxActions()
	if cmuxActionsPresent {
		if os.Getenv("CMUX_SURFACE_ID") == "" {
			fmt.Printf("  \033[33m⚠ cmux actions found but not running inside cmux — cannot remove them\033[0m\n")
		} else {
			fmt.Printf("  Remove cmux actions from %s:\n", cmuxConfigFilePath())
			for _, id := range handlerCmuxActionIDs {
				fmt.Printf("    - %s\n", id)
			}
		}
	}

	// Check watcher schedules
	for _, name := range watcherPkg.KnownWatchers {
		if watcherPkg.IsInstalled(name) {
			fmt.Printf("  Uninstall %s watcher schedule\n", name)
		}
	}

	// Detect the binary location
	binaryPath, _ := os.Executable()
	realBinaryPath, _ := filepath.EvalSymlinks(binaryPath)
	if realBinaryPath != "" {
		if binaryPath != realBinaryPath {
			fmt.Printf("  Remove %s -> %s\n", binaryPath, realBinaryPath)
		} else {
			fmt.Printf("  Remove %s\n", realBinaryPath)
		}
	}

	hooksPath := filepath.Join(agentHandlerDir, "hooks")
	skillsPath := filepath.Join(agentHandlerDir, "skills")
	rulesPath := filepath.Join(agentHandlerDir, "rules")
	if _, err := os.Stat(hooksPath); err == nil {
		fmt.Printf("  Remove extracted hooks from %s\n", hooksPath)
	}
	if _, err := os.Stat(skillsPath); err == nil {
		fmt.Printf("  Remove extracted skills from %s\n", skillsPath)
	}
	if _, err := os.Stat(rulesPath); err == nil {
		fmt.Printf("  Remove extracted rules from %s\n", rulesPath)
	}

	// Check shell completions
	completionTargets := allCompletionTargets()
	var existingCompletions []completionTarget
	for _, ct := range completionTargets {
		if _, err := os.Stat(ct.path); err == nil {
			existingCompletions = append(existingCompletions, ct)
		}
	}
	if len(existingCompletions) > 0 {
		fmt.Printf("  Remove shell completions:\n")
		for _, ct := range existingCompletions {
			fmt.Printf("    - %s\n", ct.path)
		}
	}

	fmt.Println("")

	if !confirm("Proceed?") {
		fmt.Println("Aborted.")
		return nil
	}
	fmt.Println("")

	for _, pd := range perDir {
		for _, name := range pd.symlinkTargets {
			dst := filepath.Join(pd.claudeSkillsDir, name)
			os.Remove(dst)
			fmt.Printf("  ✓ Removed skill symlink %s (%s)\n", name, pd.dir)
		}
	}

	// Remove cmux actions (only if inside cmux)
	if cmuxActionsPresent && os.Getenv("CMUX_SURFACE_ID") != "" {
		removeCmuxActions()
	}

	// Uninstall watcher schedules
	for _, name := range watcherPkg.KnownWatchers {
		if watcherPkg.IsInstalled(name) {
			watcherPkg.Uninstall(name)
			fmt.Printf("  ✓ Uninstalled %s watcher\n", name)
		}
	}

	for _, pd := range perDir {
		// Remove installed global rules
		for _, name := range pd.installedRules {
			os.Remove(filepath.Join(pd.claudeRulesDir, name))
			fmt.Printf("  ✓ Removed global rule %s (%s)\n", name, pd.dir)
		}

		if len(pd.hookNames) > 0 {
			removeHooks(pd.dir)
		}
	}

	// Remove shell completions
	removeAllCompletions()

	// Remove binary last (since we're running from it)
	if realBinaryPath != "" {
		if binaryPath != realBinaryPath {
			os.Remove(binaryPath)
			fmt.Printf("  ✓ Removed %s\n", binaryPath)
		}
		os.Remove(realBinaryPath)
		fmt.Printf("  ✓ Removed %s\n", realBinaryPath)
	}

	// Remove extracted hooks and skills from ~/.agent-handler
	for _, dir := range []string{"hooks", "skills", "rules"} {
		dirPath := filepath.Join(agentHandlerDir, dir)
		if _, err := os.Stat(dirPath); err == nil {
			os.RemoveAll(dirPath)
			fmt.Printf("  ✓ Removed %s\n", dirPath)
		}
	}

	fmt.Println("\n✓ Uninstallation complete!")
	dataDir := filepath.Join(agentHandlerDir, "data")
	if _, err := os.Stat(dataDir); err == nil {
		fmt.Printf("\n  Your event history, session data, and database are still at %s\n", dataDir)
		fmt.Println("  To fully remove all data: rm -rf ~/.agent-handler")
	}
	return nil
}

func findAgentHandlerSkills(claudeSkillsDir string) []string {
	var found []string
	for _, name := range skillNames {
		dst := filepath.Join(claudeSkillsDir, name)
		info, err := os.Lstat(dst)
		if err != nil {
			continue
		}
		if info.Mode()&os.ModeSymlink != 0 {
			target, err := os.Readlink(dst)
			if err != nil {
				continue
			}
			if strings.Contains(target, "agent-handler") {
				found = append(found, name)
			}
		}
	}
	return found
}

func findAgentHandlerHooks(settingsPath string) []string {
	data, err := os.ReadFile(settingsPath)
	if err != nil {
		return nil
	}
	settings := make(map[string]interface{})
	if err := json.Unmarshal(data, &settings); err != nil {
		return nil
	}
	hooks, ok := settings["hooks"].(map[string]interface{})
	if !ok {
		return nil
	}

	var found []string
	for _, event := range []string{"SessionEnd", "UserPromptSubmit", "PreCompact"} {
		existing, ok := hooks[event].([]interface{})
		if !ok {
			continue
		}
		for _, group := range existing {
			if isAgentHandlerHook(group) {
				found = append(found, event)
				break
			}
		}
	}
	return found
}

func removeHooks(claudeDir string) error {
	settingsPath := filepath.Join(claudeDir, "settings.json")

	data, err := os.ReadFile(settingsPath)
	if err != nil {
		return err
	}

	settings := make(map[string]interface{})
	if err := json.Unmarshal(data, &settings); err != nil {
		return err
	}

	hooks, ok := settings["hooks"].(map[string]interface{})
	if !ok {
		return nil
	}

	for _, event := range []string{"SessionStart", "SessionEnd", "UserPromptSubmit", "PreCompact"} {
		existing, ok := hooks[event].([]interface{})
		if !ok {
			continue
		}
		var kept []interface{}
		removed := false
		for _, group := range existing {
			if isAgentHandlerHook(group) {
				removed = true
			} else {
				kept = append(kept, group)
			}
		}
		if removed {
			if len(kept) == 0 {
				delete(hooks, event)
			} else {
				hooks[event] = kept
			}
			fmt.Printf("  ✓ Removed %s hook\n", event)
		}
	}

	if len(hooks) == 0 {
		delete(settings, "hooks")
	} else {
		settings["hooks"] = hooks
	}

	// Remove status line if it's ours
	if sl, ok := settings["statusLine"]; ok {
		if isAgentHandlerHook(sl) {
			delete(settings, "statusLine")
			fmt.Println("  ✓ Removed status line configuration")
		}
	}

	// Remove handler permission
	if perms, ok := settings["permissions"].(map[string]interface{}); ok {
		if allow, ok := perms["allow"].([]interface{}); ok {
			var kept []interface{}
			for _, p := range allow {
				if s, ok := p.(string); ok && s == "Bash(handler *)" {
					continue
				}
				kept = append(kept, p)
			}
			if len(kept) != len(allow) {
				perms["allow"] = kept
				fmt.Println("  ✓ Removed Bash(handler *) permission")
			}
		}
	}

	out, err := json.MarshalIndent(settings, "", "  ")
	if err != nil {
		return err
	}
	out = append(out, '\n')
	return os.WriteFile(settingsPath, out, 0644)
}

func hasHandlerPermission(settingsPath string) bool {
	data, err := os.ReadFile(settingsPath)
	if err != nil {
		return false
	}
	settings := make(map[string]interface{})
	if err := json.Unmarshal(data, &settings); err != nil {
		return false
	}
	perms, _ := settings["permissions"].(map[string]interface{})
	if perms == nil {
		return false
	}
	allow, _ := perms["allow"].([]interface{})
	for _, p := range allow {
		if s, ok := p.(string); ok && s == "Bash(handler *)" {
			return true
		}
	}
	return false
}

func isAgentHandlerHook(hookConfig interface{}) bool {
	data, err := json.Marshal(hookConfig)
	if err != nil {
		return false
	}
	s := string(data)
	return strings.Contains(s, "agent-handler")
}

func confirm(prompt string) bool {
	fmt.Printf("%s [y/N] ", prompt)
	reader := bufio.NewReader(os.Stdin)
	response, err := reader.ReadString('\n')
	if err != nil {
		return false
	}
	response = strings.TrimSpace(strings.ToLower(response))
	return response == "y" || response == "yes"
}

File: src/Handler/Cmd/Uninstall.hs
-- | Port of cmd/uninstall.go: remove agent-handler configuration.
module Handler.Cmd.Uninstall
  ( uninstallCommand
  , skillNames
  , confirm
  , isAgentHandlerHook
  ) where

import Control.Exception (IOException, try)
import Control.Monad (filterM, forM_, unless, when)
import qualified Data.Aeson as A
import qualified Data.Aeson.Key as Key
import qualified Data.Aeson.KeyMap as KM
import Data.Aeson.Encode.Pretty (Config(..), Indent(..), defConfig, encodePretty')
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Char8 as BLC
import Data.Maybe (fromMaybe, mapMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Data.Vector as V
import Options.Applicative
import System.Directory
  ( canonicalizePath, doesDirectoryExist, doesFileExist, getHomeDirectory
  , listDirectory, pathIsSymbolicLink, removeDirectoryRecursive, removeFile
  , getSymbolicLinkTarget )
import System.Environment (getExecutablePath, lookupEnv)
import System.FilePath (takeFileName, (</>))
import System.IO (hFlush, stdout)

import Handler.Cli.Common
import Handler.CmuxConfig (cmuxConfigFilePath, handlerCmuxActionIDs, hasCmuxActions, removeCmuxActions)
import Handler.Db (handlerHome)
import Handler.Watcher.Scheduler (isInstalled, uninstallWatcher)

-- | Skills installed by handler setup. Update when adding/removing skills.
skillNames :: [Text]
skillNames =
  [ "inbox"
  , "inbox-clear"
  , "inbox-mode"
  , "message"
  , "watch"
  , "unwatch"
  , "watching"
  , "handler"
  , "catchup"
  , "done"
  , "handler-debug"
  ]

-- | y/N confirmation prompt on stdin.
confirm :: Text -> IO Bool
confirm prompt = do
  printfT' [prompt, " [y/N] "]
  hFlush stdout
  result <- try getLine :: IO (Either IOException String)
  pure $ case result of
    Left _ -> False
    Right response ->
      let r = T.toLower (T.strip (T.pack response))
      in r == "y" || r == "yes"
  where printfT' = putStr . T.unpack . T.concat

-- | Whether a settings.json hook (or statusLine) entry belongs to
-- agent-handler, judged by its JSON serialization mentioning agent-handler.
isAgentHandlerHook :: A.Value -> Bool
isAgentHandlerHook v = "agent-handler" `T.isInfixOf` TE.decodeUtf8Lenient (BLC.toStrict (A.encode v))

uninstallCommand :: Mod CommandFields NamedCommand
uninstallCommand = mkCommand "uninstall" "Remove agent-handler configuration" (pure runUninstall)

runUninstall :: Ctx -> IO ()
runUninstall _ = do
  home <- getHomeDirectory
  agentHandlerDir <- handlerHome
  let claudeSkillsDir = home </> ".claude" </> "skills"
      claudeRulesDir = home </> ".claude" </> "rules"
      settingsPath = home </> ".claude" </> "settings.json"
      hooksPath = agentHandlerDir </> "hooks"
      skillsPath = agentHandlerDir </> "skills"
      rulesPath = agentHandlerDir </> "rules"

  putTextLn "agent-handler uninstall will:"
  putTextLn ""

  symlinkTargets <- findAgentHandlerSkills claudeSkillsDir
  unless (null symlinkTargets) $ do
    printfT ["  Remove ", T.pack (show (length symlinkTargets)), " skill symlinks from ", T.pack claudeSkillsDir, ":"]
    forM_ symlinkTargets $ \name -> printfT ["    - ", name]

  hookNames <- findAgentHandlerHooks settingsPath
  unless (null hookNames) $ do
    printfT ["  Remove ", T.pack (show (length hookNames)), " hooks from ", T.pack settingsPath, ":"]
    forM_ hookNames $ \name -> printfT ["    - ", name]

  cmuxActionsPresent <- hasCmuxActions
  insideCmux <- (/= Nothing) <$> lookupEnv "CMUX_SURFACE_ID"
  when cmuxActionsPresent $
    if not insideCmux
      then putTextLn "  \ESC[33m⚠ cmux actions found but not running inside cmux — cannot remove them\ESC[0m"
      else do
        cmuxPath <- cmuxConfigFilePath
        printfT ["  Remove cmux actions from ", T.pack cmuxPath, ":"]
        forM_ handlerCmuxActionIDs $ \actionId -> printfT ["    - ", actionId]

  installedWatchers <- filterM isInstalled ["github", "jira"]
  forM_ installedWatchers $ \name ->
    printfT ["  Uninstall ", name, " watcher schedule"]

  binaryPath <- getExecutablePath
  realBinaryPath <- fromMaybe binaryPath . either (const Nothing) Just
    <$> (try (canonicalizePath binaryPath) :: IO (Either IOException FilePath))
  if binaryPath /= realBinaryPath
    then printfT ["  Remove ", T.pack binaryPath, " -> ", T.pack realBinaryPath]
    else printfT ["  Remove ", T.pack realBinaryPath]

  forM_ [(hooksPath, "hooks"), (skillsPath, "skills"), (rulesPath, "rules")] $ \(p, what) -> do
    exists <- doesDirectoryExist p
    when exists $ printfT ["  Remove extracted ", what, " from ", T.pack p]

  installedRules <- do
    entries <- either (const []) id
      <$> (try (listDirectory rulesPath) :: IO (Either IOException [FilePath]))
    filterM (\name -> doesFileExist (claudeRulesDir </> name)) entries
  unless (null installedRules) $ do
    printfT ["  Remove ", T.pack (show (length installedRules)), " global rule(s) from ", T.pack claudeRulesDir, ":"]
    forM_ installedRules $ \name -> printfT ["    - ", T.pack name]

  hasPerm <- hasHandlerPermission settingsPath
  when hasPerm $
    printfT ["  Remove Bash(handler *) permission from ", T.pack settingsPath]

  putTextLn ""
  proceed <- confirm "Proceed?"
  if not proceed
    then putTextLn "Aborted."
    else do
      putTextLn ""

      forM_ symlinkTargets $ \name -> do
        _ <- try (removeFile (claudeSkillsDir </> T.unpack name)) :: IO (Either IOException ())
        printfT ["  ✓ Removed skill symlink ", name]

      when (cmuxActionsPresent && insideCmux) removeCmuxActions

      forM_ installedWatchers $ \name -> do
        uninstallWatcher name
        printfT ["  ✓ Uninstalled ", name, " watcher"]

      forM_ installedRules $ \name -> do
        _ <- try (removeFile (claudeRulesDir </> name)) :: IO (Either IOException ())
        printfT ["  ✓ Removed global rule ", T.pack name]

      unless (null hookNames) $ removeHooks home

      -- Remove binary last (since we're running from it)
      when (binaryPath /= realBinaryPath) $ do
        _ <- try (removeFile binaryPath) :: IO (Either IOException ())
        printfT ["  ✓ Removed ", T.pack binaryPath]
      _ <- try (removeFile realBinaryPath) :: IO (Either IOException ())
      printfT ["  ✓ Removed ", T.pack realBinaryPath]

      forM_ ["hooks", "skills", "rules"] $ \dir -> do
        let dirPath = agentHandlerDir </> dir
        exists <- doesDirectoryExist dirPath
        when exists $ do
          removeDirectoryRecursive dirPath
          printfT ["  ✓ Removed ", T.pack dirPath]

      putTextLn "\n✓ Uninstallation complete!"
      let dataDir = agentHandlerDir </> "data"
      dataExists <- doesDirectoryExist dataDir
      when dataExists $ do
        printfT ["\n  Your event history, session data, and database are still at ", T.pack dataDir]
        putTextLn "  To fully remove all data: rm -rf ~/.agent-handler"

-- | Skill symlinks in ~\/.claude\/skills that point into agent-handler.
findAgentHandlerSkills :: FilePath -> IO [Text]
findAgentHandlerSkills claudeSkillsDir =
  flip filterM skillNames $ \name -> do
    let dst = claudeSkillsDir </> T.unpack name
    isLink <- either (const False) id
      <$> (try (pathIsSymbolicLink dst) :: IO (Either IOException Bool))
    if not isLink
      then pure False
      else do
        target <- try (getSymbolicLinkTarget dst) :: IO (Either IOException FilePath)
        pure $ case target of
          Right t -> "agent-handler" `T.isInfixOf` T.pack t
          Left _  -> False

readSettings :: FilePath -> IO (Maybe (KM.KeyMap A.Value))
readSettings settingsPath = do
  result <- try (BL.readFile settingsPath) :: IO (Either IOException BL.ByteString)
  pure $ case result of
    Left _ -> Nothing
    Right raw -> case A.decode raw of
      Just (A.Object o) -> Just o
      _ -> Nothing

writeSettings :: FilePath -> KM.KeyMap A.Value -> IO ()
writeSettings settingsPath settings =
  BL.writeFile settingsPath (encodePretty' cfg (A.Object settings) <> "\n")
  where cfg = defConfig { confIndent = Spaces 2, confCompare = compare }

hookEvents :: [Text]
hookEvents = ["SessionEnd", "UserPromptSubmit", "PreCompact"]

-- | Hook events in settings.json that contain agent-handler matcher groups.
findAgentHandlerHooks :: FilePath -> IO [Text]
findAgentHandlerHooks settingsPath = do
  msettings <- readSettings settingsPath
  pure $ case msettings of
    Nothing -> []
    Just settings -> case KM.lookup "hooks" settings of
      Just (A.Object hooks) ->
        [ event
        | event <- hookEvents
        , Just (A.Array groups) <- [KM.lookup (Key.fromText event) hooks]
        , any isAgentHandlerHook (V.toList groups)
        ]
      _ -> []

-- | Removes agent-handler hook groups, statusLine, and the Bash(handler *)
-- permission from settings.json.
removeHooks :: FilePath -> IO ()
removeHooks home = do
  let settingsPath = home </> ".claude" </> "settings.json"
  msettings <- readSettings settingsPath
  case msettings of
    Nothing -> pure ()
    Just settings0 -> do
      let hooks0 = case KM.lookup "hooks" settings0 of
            Just (A.Object h) -> h
            _ -> KM.empty
      hooks1 <- foldHooks hooks0 ["SessionStart", "SessionEnd", "UserPromptSubmit", "PreCompact"]
      let settings1 =
            if KM.null hooks1
              then KM.delete "hooks" settings0
              else KM.insert "hooks" (A.Object hooks1) settings0

      settings2 <- case KM.lookup "statusLine" settings1 of
        Just sl | isAgentHandlerHook sl -> do
          putTextLn "  ✓ Removed status line configuration"
          pure (KM.delete "statusLine" settings1)
        _ -> pure settings1

      settings3 <- case KM.lookup "permissions" settings2 of
        Just (A.Object perms) | Just (A.Array allow) <- KM.lookup "allow" perms -> do
          let kept = V.filter (/= A.String "Bash(handler *)") allow
          if V.length kept /= V.length allow
            then do
              putTextLn "  ✓ Removed Bash(handler *) permission"
              pure (KM.insert "permissions"
                     (A.Object (KM.insert "allow" (A.Array kept) perms)) settings2)
            else pure settings2
        _ -> pure settings2

      writeSettings settingsPath settings3
  where
    foldHooks hooks [] = pure hooks
    foldHooks hooks (event : rest) =
      case KM.lookup (Key.fromText event) hooks of
        Just (A.Array groups) -> do
          let kept = V.filter (not . isAgentHandlerHook) groups
          if V.length kept /= V.length groups
            then do
              printfT ["  ✓ Removed ", event, " hook"]
              let hooks' = if V.null kept
                    then KM.delete (Key.fromText event) hooks
                    else KM.insert (Key.fromText event) (A.Array kept) hooks
              foldHooks hooks' rest
            else foldHooks hooks rest
        _ -> foldHooks hooks rest

-- | Whether Bash(handler *) is in the settings.json allow list.
hasHandlerPermission :: FilePath -> IO Bool
hasHandlerPermission settingsPath = do
  msettings <- readSettings settingsPath
  pure $ case msettings of
    Just settings
      | Just (A.Object perms) <- KM.lookup "permissions" settings
      , Just (A.Array allow) <- KM.lookup "allow" perms
      -> V.elem (A.String "Bash(handler *)") allow
    _ -> False
